src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java

```diff
@@ -227,11 +239,44 @@ protected PathFragment readSymbolicLink(PathFragment path) throws IOException {
   @Override
   protected void createSymbolicLink(PathFragment linkPath, PathFragment targetFragment)
       throws IOException {
-    /*
-     * TODO(buchgr): Optimize the case where we are creating a symlink to a remote output. This does
-     * add a non-trivial amount of complications though (as symlinks tend to do).
-     */
-    downloadFileIfRemote(execRoot.getRelative(targetFragment));
+    FileStatus stat = statNullable(linkPath, /* followSymlinks= */ false);
+    if (stat != null) {
+      throw new IOException(linkPath + " (File exists)");
+    }
+
+    // If the output is a regular file or tree artifact, it has "copy" semantics, i.e., it will be
+    // transparently resolved into the artifact it points to; the symlink is just a filesystem-level
+    // optimization to avoid making an actual copy. We can inject its metadata and avoid the need to
+    // compute it from the filesystem later, which would be slower and require additional filesystem
+    // operations to be implemented.
+    Artifact output = outputArtifactByExecPath.get(linkPath.relativeTo(execRoot));
+    if (output != null) {
+      if (output.isTreeArtifact()) {
+        TreeArtifactValue metadata = getRemoteInputTreeMetadata(targetFragment);
+        if (metadata != null) {
+          // The archived representation is a Google-internal feature disabled on open source Bazel.
+          checkState(
+              metadata.getArchivedRepresentation().isEmpty(),
+              "symlink to archived tree artifact is not supported");
+          TreeArtifactValue.Builder treeBuilder =
+              TreeArtifactValue.newBuilder((SpecialArtifact) output);
+          for (Map.Entry<TreeFileArtifact, FileArtifactValue> entry :
+              metadata.getChildValues().entrySet()) {
+            treeBuilder.putChild(
+                TreeFileArtifact.createTreeOutput(
+                    (SpecialArtifact) output, entry.getKey().getParentRelativePath()),
+                entry.getValue());
+          }
+          injectTree((SpecialArtifact) output, treeBuilder.build());
+        }
+      } else if (!output.isSymlink()) {
+        RemoteFileArtifactValue metadata = getRemoteInputMetadata(targetFragment);
+        if (metadata != null) {
+          injectFile(output, metadata);
+        }
+      }
```

Comment:
What about output that is symlink? I know we are going to handle it differently, but should we throw an error here?

Reply:
If the output is a symlink, then we should create the symlink and not inject any metadata. Skyframe will later call `readSymbolicLink` and not attempt to resolve it.

Unfortunately, unresolved symlinks don't currently work together with remote execution because of other issues, so I can't add an integration test for it (I did add a unit test to verify that `createSymbolicLink` and `readSymbolicLink` round-trip when the output is of type symlink).

Reply:
That makes sense. Do we allow actions to access undeclared output (i.e. create undeclared symlink in this case)?

Reply:
Yes, undeclared symlinks are created via `ctx.actions.symlink(output = declare_symlink(...))` which works today. They can also be consumed as inputs to other actions.

I suspect we should also inject the metadata for undeclared symlinks into Skyframe (`UndeclaredSymlinkFileArtifactValue` seems to have been introduced for that exact purpose, but is mysteriously unused in the codebase). However, I don't want to expand the scope of this PR to include it, so for now I'm still creating an actual symlink in the local filesystem.

Reply:
Sorry for didn't make it clear, but my question was about **undeclared** output not **unresolved** symlink. To be more specific, do we want to move line `super.createSymbolicLink()` inside the check `if (output != null)` and instead throw an error there to prevent actions create **undeclared** outputs?

Reply:
I think it's reasonable to throw if the symlink doesn't point to an output, yes. This filesystem is only meant to be used by internal actions, and the set of things they can do is fairly restricted.